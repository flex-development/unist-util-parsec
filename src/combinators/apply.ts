/**
 * @file Combinators - apply
 * @module unist-util-parsec/combinators/apply
 */

import type { Parser, Token } from '#src/interfaces'
import type {
  ApplyCallback,
  ParserOutput,
  TokenRange,
  TokenType
} from '#src/types'

/**
 * Parse an occurrence of `x`, and transform the result with `f`.
 *
 * Fails if `x` fails.
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode Parser}
 * @see {@linkcode TokenRange}
 * @see {@linkcode TokenType}
 *
 * @template {TokenType} T - Token type
 * @template {any} From - Original parse candidate result
 * @template {any} To - Final parse candidate result
 * @template {TokenRange} [Range=TokenRange] - Token range
 *
 * @param {Parser<T, From>} x - Parser to apply
 * @param {ApplyCallback<From, To, Range>} f - Transform callback
 * @return {Parser<T, To>} Applicative parser
 */
function apply<
  T extends TokenType,
  From,
  To,
  Range extends TokenRange = TokenRange
>(x: Parser<T, From>, f: ApplyCallback<From, To, Range>): Parser<T, To> {
  return { parse }

  /**
   * Consume `token` and transform the result.
   *
   * @param {Token<T>?} [token] - Current token
   * @return {ParserOutput<T, To>} Parser output
   */
  function parse(token?: Token<T>): ParserOutput<T, To> {
    /**
     * Parser output.
     *
     * @const {ParserOutput<T, V>} output
     */
    const output: ParserOutput<T, From> = x.parse(token)

    // transform parse candidate result
    if (output.successful) {
      return {
        candidate: {
          head: token,
          next: output.candidate.next,
          result: f(output.candidate.result, [
            token,
            output.candidate.next
          ] as unknown as Range)
        },
        error: output.error,
        successful: true
      }
    }

    return output
  }
}

export default apply
