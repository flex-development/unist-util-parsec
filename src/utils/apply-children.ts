/**
 * @file Utilities - applyChildren
 * @module unist-util-parsec/utils/applyChildren
 */

import type { Token } from '#src/interfaces'
import type { ApplyNode, InitChildren } from '#src/types'
import type { Children, Match, Type } from '@flex-development/unist-util-types'
import type { Node, Parent } from 'unist'
import isToken from './is-token'

/**
 * Create a child node array.
 *
 * > ðŸ‘‰ An array initializer can have a maximum depth of `5`.
 *
 * @see {@linkcode ApplyNode}
 * @see {@linkcode Children}
 * @see {@linkcode InitChildren}
 *
 * @template {ApplyNode} N - Parent node
 *
 * @param {InitChildren<N>} init - Initializer
 * @return {Children<N>} New child node array
 */
function applyChildren<N extends ApplyNode>(init: InitChildren<N>): Children<N>

/**
 * Create a child node array.
 *
 * > ðŸ‘‰ An array initializer can have a maximum depth of `5`.
 *
 * @see {@linkcode Children}
 * @see {@linkcode InitChildren}
 * @see {@linkcode Node}
 *
 * @template {Node} N - Parent node
 *
 * @param {InitChildren<N>} init - Initializer
 * @return {Children<N>} New child node array
 */
function applyChildren<N extends Node>(init: InitChildren<N>): Children<N>

/**
 * Create a child node array.
 *
 * > ðŸ‘‰ An array initializer can have a maximum depth of `5`.
 *
 * @see {@linkcode ApplyNode}
 * @see {@linkcode Children}
 * @see {@linkcode InitChildren}
 * @see {@linkcode Match}
 * @see {@linkcode Type}
 *
 * @template {Type<ApplyNode>} T - Node type
 *
 * @param {InitChildren<Match<ApplyNode, T>>} init - Initializer
 * @return {Children<Match<ApplyNode, T>>} New child node array
 */
function applyChildren<
  T extends Type<ApplyNode>
>(init: InitChildren<Match<ApplyNode, T>>): Children<Match<ApplyNode, T>>

/**
 * Create a child node array.
 *
 * > ðŸ‘‰ An array initializer can have a maximum depth of `5`.
 *
 * @see {@linkcode Children}
 * @see {@linkcode InitChildren}
 * @see {@linkcode Parent}
 *
 * @param {InitChildren} init - Initializer
 * @return {Children<Parent>} New child node array
 */
function applyChildren(init: InitChildren): Children<Parent> {
  return [init].flat(6).filter((value?: Node | Token | null): value is Node => {
    return !!value && !isToken(value)
  })
}

export default applyChildren
