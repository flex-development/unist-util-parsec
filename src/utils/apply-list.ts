/**
 * @file Utilities - applyList
 * @module unist-util-parsec/utils/applyList
 */

import type { ListResult } from '#src/types'

/**
 * Create a list result.
 *
 * @see {@linkcode ListResult}
 *
 * @template {any} [R=any] - List item parse candidate result
 *
 * @param {[R, R[]]} value - Apply callback value
 * @return {ListResult<R>} List result
 */
function applyList<R = any>(value: [R, R[]]): ListResult<R> {
  return [value[0], ...value[1]]
}

export default applyList
