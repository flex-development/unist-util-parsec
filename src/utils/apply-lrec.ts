/**
 * @file Utilities - applyLrec
 * @module unist-util-parsec/utils/applyLrec
 */

import type {
  ApplyCallback,
  LrecCallback,
  RepResult,
  TokenRange
} from '#src/types'

/**
 * Create an applicative parser callback for a left recursive parser.
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode LrecCallback}
 *
 * @template {any} [R=any] - Final parse candidate result
 * @template {any} [P=any] - First parse candidate result
 * @template {any} [Q=any] - Second parse candidate result
 *
 * @param {LrecCallback<R, P, Q>} f - Parse candidate result callback
 * @return {ApplyCallback<[P, RepResult<[Q, TokenRange]>], R>} Apply callback
 */
function applyLrec<R = any, P = any, Q = any>(
  f: LrecCallback<R, P, Q>
): ApplyCallback<[P, RepResult<[Q, TokenRange]>], R> {
  return callback

  /**
   * Create a recursive, left-associative parse candidate result.
   *
   * @param {[P, RepResult<[Q, TokenRange]>]} value - Apply callback value
   * @return {R} Final parse candidate result
   */
  function callback(value: [P, RepResult<[Q, TokenRange]>]): R {
    /**
     * Parse candidate result.
     *
     * @var {P | R} result
     */
    let result: P | R = value[0]

    for (const tail of value[1]) result = f(result, ...tail)

    return <R>result
  }
}

export default applyLrec
