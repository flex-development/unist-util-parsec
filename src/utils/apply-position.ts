/**
 * @file Utilities - applyPosition
 * @module unist-util-parsec/utils/applyPosition
 */

import type { TokenRange } from '#src/types'
import type { Node, Parent } from 'unist'

/**
 * Set the position of `node` based on token `range` or its children.
 *
 * > ðŸ‘‰ Does nothing under the following conditions:
 * >
 * > - `range` does not include two tokens (range based position)
 * > - `node` is not a [*parent*][parent] (children-based position)
 * > - the [*head*][head] or [*tail*][tail] of `node` is
 * >   [*generated*][generated] (children-based position)
 *
 * [generated]: https://github.com/syntax-tree/unist#generated
 * [head]: https://github.com/syntax-tree/unist#head
 * [parent]: https://github.com/syntax-tree/unist#parent
 * [tail]: https://github.com/syntax-tree/unist#tail
 *
 * @see {@linkcode Node}
 * @see {@linkcode Parent}
 * @see {@linkcode TokenRange}
 *
 * @template {Node | Parent} T - Node to position
 *
 * @param {T} node - Node to position
 * @param {(TokenRange | null)?} [range] - Token range of `node`
 * @param {(boolean | null)?} [reverse] - Use reverse positioning logic when
 * setting position based on token range
 * @return {T} `node`
 */
function applyPosition<T extends Node | Parent>(
  node: T,
  range?: TokenRange | null,
  reverse?: boolean | null
): T {
  if (!node.position) {
    if (range) {
      const [head, tail] = range

      if (head && tail) {
        if (reverse && head.previous) {
          node.position = { end: tail.start, start: head.previous.end }
        } else if (tail.previous) {
          node.position = { end: tail.previous.end, start: head.start }
        }
      }

      if (!node.position && 'children' in node) node = applyPosition(node)
    } else if ('children' in node) {
      /**
       * Head of node.
       *
       * @const {Node | undefined} head
       */
      const head: Node | undefined = node.children[0]

      /**
       * Tail of node.
       *
       * @const {Node | undefined} tail
       */
      const tail: Node | undefined = node.children.at(-1)

      if (head && tail && head.position && tail.position) {
        node.position = { end: tail.position.end, start: head.position.start }
      }
    }
  }

  return node
}

export default applyPosition
