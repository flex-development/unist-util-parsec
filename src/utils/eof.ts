/**
 * @file Utilities - eof
 * @module unist-util-parsec/utils/eof
 */

import { ParseError } from '#src/errors'
import type { ParseCandidate } from '#src/interfaces'
import type { ParserOutput, TokenType } from '#src/types'
import out from './out'

/**
 * Ensure all tokens are consumed.
 *
 * @see {@linkcode ParseCandidate}
 * @see {@linkcode ParserOutput}
 * @see {@linkcode TokenType}
 *
 * @template {TokenType} T - Token type
 * @template {any} R - Parse candidate result
 *
 * @param {ParserOutput<T, R>} output - Parser output
 * @return {typeof output} `output`
 */
function eof<T extends TokenType, R>(
  output: ParserOutput<T, R>
): typeof output {
  if (!output.successful) return output

  /**
   * Parse error.
   *
   * @var {ParseError | null | undefined} error
   */
  let error: ParseError | null | undefined = output.error

  // fail output on missing candidate or unconsumed token
  if (!<ParseCandidate | null>output.candidate) {
    error = new ParseError(null, 'No parse candidate found')
  } else if (output.candidate.next) {
    error = new ParseError(output.candidate.next, 'Unable to reach end of file')
  }

  return out(
    !!<ParseCandidate | null>output.candidate && !output.candidate.next,
    output.candidate,
    error
  )
}

export default eof
