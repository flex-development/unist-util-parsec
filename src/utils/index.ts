/**
 * @file Entry Point - Utilities
 * @module unist-util-parsec/utils
 */

export { default as applyChildren } from './apply-children'
export { default as applyList } from './apply-list'
export { default as applyLrec } from './apply-lrec'
export { default as applyPosition } from './apply-position'
export { default as eof } from './eof'
export { default as nok } from './nok'
export { default as out } from './out'
export { default as result } from './result'
