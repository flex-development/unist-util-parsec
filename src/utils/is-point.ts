/**
 * @file Utilities - isPoint
 * @module unist-util-parsec/utils/isPoint
 */

import type { Point } from '#src/interfaces'

/**
 * Check if the specified `value` is a point.
 *
 * @see {@linkcode Point}
 *
 * @internal
 *
 * @param {unknown} value - Value to check
 * @return {value is Point} `true` if `value` is point
 */
function isPoint(value: unknown): value is Point {
  return (
    typeof value === 'object' &&
    value !== null &&
    'column' in value &&
    'line' in value &&
    'offset' in value &&
    typeof value.column === 'number' &&
    typeof value.line === 'number' &&
    typeof value.offset === 'number'
  )
}

export default isPoint
