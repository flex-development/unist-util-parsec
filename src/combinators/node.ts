/**
 * @file Combinators - node
 * @module unist-util-parsec/combinators/node
 */

import type { Parser } from '#src/interfaces'
import type {
  ApplyCallback,
  ApplyNode,
  Builder,
  InitChildren,
  TokenRange,
  TokenType as TT
} from '#src/types'
import { applyChildren, applyPosition } from '#src/utils'
import { u } from '@flex-development/unist-util-builder'
import type { Match, Type } from '@flex-development/unist-util-types'
import { ok } from 'devlop'
import type { Node } from 'unist'
import apply from './apply'

/**
 * Parse an occurrence of `p`, and transform the result with `f` to build a
 * node of the specified `type`.
 *
 * The callback receives the value of `p` and should return a child node array,
 * properties object, or value that can be passed to {@linkcode u}.
 *
 * The following types can be used as builder values:
 *
 * - `bigint`
 * - `boolean`
 * - `number`
 * - `string`
 * - `null`
 *
 * > ðŸ‘‰ Undefined literals must be created using a properties object, rather
 * > than a value. Returning `undefined` from `f` will create a void node.
 *
 * The node will be positioned using the current token range, or its children.
 *
 * Fails if `p` fails.
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode ApplyNode}
 * @see {@linkcode Builder}
 * @see {@linkcode Match}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode TokenRange}
 * @see {@linkcode Type}
 *
 * @template {Type<ApplyNode>} T - Node type
 * @template {any} [V=any] - Apply callback value
 * @template {TokenRange} [R=TokenRange] - Token range
 *
 * @param {T} type - Node type
 * @param {Parser<TT, V>} p - Parser to apply
 * @param {ApplyCallback<V, Builder<T> | undefined, R>} f - Apply callback
 * @return {Parser<TT, Match<ApplyNode, T>>} Applicative node parser
 */
function node<
  T extends Type<ApplyNode>,
  V,
  R extends TokenRange = TokenRange
>(
  type: T,
  p: Parser<TT, V>,
  f: ApplyCallback<V, Builder<T> | undefined, R>
): Parser<TT, Match<ApplyNode, T>>

/**
 * Parse an occurrence of `p`, and transform the result with `f` to build a
 * node of the specified `type`.
 *
 * The callback receives the value of `p` and should return a child node array,
 * properties object, or value that can be passed to {@linkcode u}.
 *
 * The following types can be used as builder values:
 *
 * - `bigint`
 * - `boolean`
 * - `number`
 * - `string`
 * - `null`
 *
 * > ðŸ‘‰ Undefined literals must be created using a properties object, rather
 * > than a value. Returning `undefined` from `f` will create a void node.
 *
 * The node will be positioned using the current token range, or its children.
 *
 * Fails if `p` fails.
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode Builder}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode TokenRange}
 * @see {@linkcode Type}
 *
 * @template {Node} N - Node to build
 * @template {any} V - Apply callback value
 * @template {TokenRange} R - Token range
 *
 * @param {Type<N>} type - Node type
 * @param {Parser<TT, V>} p - Parser to apply
 * @param {ApplyCallback<V, Builder<Type<N>> | undefined, R>} f - Apply callback
 * @return {Parser<TT, N>} Applicative node parser
 */
function node<N extends Node, V = any, R extends TokenRange = TokenRange>(
  type: Type<N>,
  p: Parser<TT, V>,
  f: ApplyCallback<V, Builder<Type<N>> | undefined, R>
): Parser<TT, N>

/**
 * Parse an occurrence of `p`, and use the result to build a parent node of the
 * specified `type`.
 *
 * > ðŸ‘‰ The value of `p` should be compatible with {@linkcode applyChildren}.
 *
 * The node will be positioned using the current token range, or its children.
 *
 * Fails if `p` fails.
 *
 * @see {@linkcode ApplyNode}
 * @see {@linkcode InitChildren}
 * @see {@linkcode Match}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode Type}
 *
 * @template {Type<ApplyNode>} T - Node children
 * @template {InitChildren<Match<ApplyNode, T>>} V - Parent node builder
 *
 * @param {Type<N>} type - Node type
 * @param {Parser<TT, V>} p - Parser to apply
 * @return {Parser<TT, Match<ApplyNode, T>>} Applicative node parser
 */
function node<
  T extends Type<ApplyNode>,
  V extends InitChildren<Match<ApplyNode, T>>
>(type: T, p: Parser<TT, V>): Parser<TT, Match<ApplyNode, T>>

/**
 * Parse an occurrence of `p`, and use the result to build a parent node of the
 * specified `type`.
 *
 * > ðŸ‘‰ The value of `p` should be compatible with {@linkcode applyChildren}.
 *
 * The node will be positioned using the current token range, or its children.
 *
 * Fails if `p` fails.
 *
 * @see {@linkcode InitChildren}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode Type}
 *
 * @template {Node} N - Node to build
 * @template {InitChildren<N>} [V=InitChildren<N>] - Apply callback value
 *
 * @param {Type<N>} type - Node type
 * @param {Parser<TT, V>} p - Parser to apply
 * @return {Parser<TT, N>} Applicative node parser
 */
function node<N extends Node, V extends InitChildren<N> = InitChildren<N>>(
  type: Type<N>,
  p: Parser<TT, V>
): Parser<TT, N>

/**
 * Parse an occurrence of `p`, and transform the result with `f`.
 *
 * The callback should return a new node. If not positioned, the new node will
 * be [*generated*][generated].
 *
 * Fails if `p` fails.
 *
 * [generated]: https://github.com/syntax-tree/unist#generated
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode ApplyNode}
 * @see {@linkcode Parser}
 * @see {@linkcode TokenRange}
 *
 * @template {ApplyNode} N - Node to build
 * @template {any} [V=any] - Apply callback value
 * @template {TokenRange} [R=TokenRange] - Token range
 *
 * @param {Parser<TT, V>} p - Parser to apply
 * @param {ApplyCallback<V, N, R>} f - Apply callback
 * @return {Parser<TT, N>} Applicative node parser
 */
function node<N extends ApplyNode, V = any, R extends TokenRange = TokenRange>(
  p: Parser<TT, V>,
  f: ApplyCallback<V, N, R>
): Parser<TT, N>

/**
 * Parse an occurrence of `p`, and transform the result with `f`.
 *
 * The callback should return a new node. If not positioned, the new node will
 * be [*generated*][generated].
 *
 * Fails if `p` fails.
 *
 * [generated]: https://github.com/syntax-tree/unist#generated
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode TokenRange}
 *
 * @template {Node} N - Node to build
 * @template {any} [V=any] - Apply callback value
 * @template {TokenRange} [R=TokenRange] - Token range
 *
 * @param {Parser<TT, V>} p - Parser to apply
 * @param {ApplyCallback<V, N, R>} f - Apply callback
 * @return {Parser<TT, N>} Applicative node parser
 */
function node<N extends Node, V = any, R extends TokenRange = TokenRange>(
  p: Parser<TT, V>,
  f: ApplyCallback<V, N, R>
): Parser<TT, N>

/**
 * Build a node.
 *
 * @see {@linkcode ApplyCallback}
 * @see {@linkcode Builder}
 * @see {@linkcode Node}
 * @see {@linkcode Parser}
 * @see {@linkcode Type}
 *
 * @param {Parser | Type} x - Parser to apply or node type
 * @param {ApplyCallback<any, Node> | Parser} p - Apply callback or parser
 * @param {ApplyCallback<any, Builder | undefined>?} f - Apply callback
 * @return {Parser<TT, Node>} Applicative node parser
 */
function node(
  x: Parser | Type,
  p: ApplyCallback<any, Node> | Parser,
  f?: ApplyCallback<any, Builder | undefined>
): Parser<TT, Node> {
  if (typeof x === 'object' && typeof p === 'function') return apply(x, p)

  ok(typeof x === 'string', 'expected node type')
  ok(typeof p === 'object' && 'parse' in p, 'expected parser')

  return apply<TT, any, Node>(p, (value, range) => {
    return applyPosition(
      typeof f === 'function'
        ? u(x, <never>f(value, range))
        : u(x, applyChildren(<never>value)),
      range
    )
  })
}

export default node
