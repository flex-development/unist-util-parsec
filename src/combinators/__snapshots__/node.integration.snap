// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:combinators/node > parser sample 0 1`] = `
root[1] (1:1-1:34, 0-33)
└─0 expressionStatement[1] (1:1-1:34, 0-33)
    └─0 arithmeticExpression[2] (1:1-1:34, 0-33)
        │ operator: "*"
        ├─0 arithmeticExpression[2] (1:1-1:27, 0-26)
        │   │ operator: "/"
        │   ├─0 parenthesizedExpression[1] (1:1-1:21, 0-20)
        │   │   └─0 arithmeticExpression[2] (1:2-1:20, 1-19)
        │   │       │ operator: "+"
        │   │       ├─0 arithmeticExpression[2] (1:2-1:9, 1-8)
        │   │       │   │ operator: "-"
        │   │       │   ├─0 number 3.5 (1:2-1:5, 1-4)
        │   │       │   │     raw: "3.5"
        │   │       │   └─1 number 5 (1:8-1:9, 7-8)
        │   │       │         raw: "5"
        │   │       └─1 parenthesizedExpression[1] (1:12-1:20, 11-19)
        │   │           └─0 arithmeticExpression[2] (1:13-1:19, 12-18)
        │   │               │ operator: "**"
        │   │               ├─0 number 2 (1:13-1:14, 12-13)
        │   │               │     raw: "2"
        │   │               └─1 number 2 (1:18-1:19, 17-18)
        │   │                     raw: "2"
        │   └─1 number 0.1 (1:24-1:27, 23-26)
        │         raw: ".10"
        └─1 boolean true (1:30-1:34, 29-33)
`;

exports[`integration:combinators/node > parser sample 1 1`] = `
root[1] (1:1-1:49, 0-48)
└─0 expressionStatement[1] (1:1-1:49, 0-48)
    └─0 templateLiteral[5] (1:1-1:49, 0-48)
        ├─0 templateElement "fifteen is \\\\\`" (1:2-1:15, 1-14)
        ├─1 templatePlaceholder[1] (1:15-1:23, 14-22)
        │   │ typeOnly: false
        │   └─0 arithmeticExpression[2] (1:17-1:22, 16-21)
        │       │ operator: "+"
        │       ├─0 identifier<a> (1:17-1:18, 16-17)
        │       └─1 identifier<b> (1:21-1:22, 20-21)
        ├─2 templateElement "\\\\\`, not \\\\\`" (1:23-1:33, 22-32)
        ├─3 templatePlaceholder[1] (1:33-1:45, 32-44)
        │   │ typeOnly: false
        │   └─0 arithmeticExpression[2] (1:35-1:44, 34-43)
        │       │ operator: "+"
        │       ├─0 arithmeticExpression[2] (1:35-1:40, 34-39)
        │       │   │ operator: "*"
        │       │   ├─0 number 2 (1:35-1:36, 34-35)
        │       │   │     raw: "2"
        │       │   └─1 identifier<a> (1:39-1:40, 38-39)
        │       └─1 identifier<b> (1:43-1:44, 42-43)
        └─4 templateElement "\\\\\`." (1:45-1:48, 44-47)
`;

exports[`integration:combinators/node > parser sample 2 1`] = `
root[7] (1:1-1:57, 0-56)
├─0 expressionStatement[1] (1:1-1:5, 0-4)
│   └─0 bigint 13n (1:1-1:5, 0-4)
│         raw: "0xdn"
├─1 expressionStatement[1] (1:6-1:11, 5-10)
│   └─0 boolean false (1:6-1:11, 5-10)
├─2 expressionStatement[1] (1:12-1:16, 11-15)
│   └─0 null null (1:12-1:16, 11-15)
├─3 expressionStatement[1] (1:17-1:18, 16-17)
│   └─0 number 3 (1:17-1:18, 16-17)
│         raw: "3"
├─4 expressionStatement[1] (1:19-1:39, 18-38)
│   └─0 regexp "/(?<=(?<a>\\\\w){3})f/u" (1:19-1:39, 18-38)
│         flags: "u"
│         pattern: "(?<=(?<a>\\\\w){3})f"
├─5 expressionStatement[1] (1:40-1:47, 39-46)
│   └─0 string "hello" (1:40-1:47, 39-46)
│         raw: "\\"hello\\""
└─6 expressionStatement[1] (1:48-1:57, 47-56)
    └─0 undefined undefined (1:48-1:57, 47-56)
`;
